### ISSUE TITLE
New conflicts with mypy in v1.0.3

### ISSUE THREAD

**Issue by KyleLangley280** [bwindsor/typed-config#10]
When updating to v1.0.3, mypy starts having conflict issues with setting keyword arguments for Callables/functions, for example:
```
filter = key(
    required=False, default=[], cast=lambda x: set(example_function(x))
)
```
Depending on what types you provide, the error from mypy will look something like:
```
error: No overload variant of "key" matches argument types "bool", "List[<nothing>]","Callable[[Any], Set[str]]"
note: Possible overload variants:
note:     def [T] key(*, section_name: Optional[str] = ..., key_name: Optional[str] = ..., required: Literal[False], cast: Callable[[Union[T, str]], T], default: T) -> T
note:     def [T] key(*, section_name: Optional[str] = ..., key_name: Optional[str] = ..., required: Literal[False], cast: Callable[[Union[T, str]], T], default: None = ...) -> Optional[T]
note:     <3 more similar overloads not shown, out of 5 total overloads>
```
I assume it has something to do with the type-checker implemented with version 1.x.x. Any insight is welcome 👍

**Comment by bwindsor** [comment]
From your `cast` function, it looks like you want `filter` is a `Set[str]` type. However, your default value is `[]` which is a `list`. So the first change required is to have
```python
filter = key(
    required=False, default=set(), cast=lambda x: set(example_function(x))
)
```

However, I believe mypy will still give you an error at this point, as the `default` value should match the return value of `cast`, and it can't tell which type is stored inside an empty set, so it will complain that `Set[<nothing>]` and `Set[str]` are not the same. So I think this should work for you as a full solution:
```python
from typing import Set

default_value: Set[str] = set()
filter = key(
    required=False, default=default_value cast=lambda x: set(example_function(x))
)
```

Let me know if this fixes the problem.

**Comment by KyleLangley280** [comment]
Same notes as before but produces the error:
```
error: No overload variant of "key" matches argument types "bool", "Set[str]", "Callable[[Any], Set[str]]"
 ```
 Strange that this hasn't been an issue that Mypy has picked up before this version update

**Comment by bwindsor** [comment]
Would you be able to post what you're using as `example_function` in this case?

**Comment by KyleLangley280** [comment]
The "example_function" is the `lower_comma_separated_str`
```
def comma_separated_str(text: str) -> list[str]:
    if isinstance(text, list):
        return text
    return text.split(",")


def lower_comma_separated_str(text: str) -> list[str]:
    return [i.lower() for i in comma_separated_str(text)]
   ```

**Comment by bwindsor** [comment]
Thanks for that, I now understand the reason for the error. The `cast` function is actually called with the `default` value, so if you have
```python
key(cast=int, default='3')
```
You would still end up with `3` as an `int`, since it would cast your default. So to work with `1.0.3` you cast function would need to be
```
def lower_comma_separated_str(text: Union[set[str], str]) -> list[str]:
    if isinstance(text, set):
        return list(text)
    return [i.lower() for i in comma_separated_str(text)]
```
which isn't partcularly logical.

Now that `key` has proper typing on it (since `1.0.0`), the above `key(cast=int, default='3')` won't pass any type checkers, as the type of `default` must match the return type of `cast`. Therefore I have removed this behaviour (which was also undocumented - the README clearly says that `cast` functions must accept a string, which I think is the sensible behaviour to have).

This should be available very shortly in `1.1.0`. Thanks very much for your examples highlighting this confusing behaviour! Please let me know if updating to `1.1.0` fixes it for you, or if `mypy` is continuing to complain.

**Comment by bwindsor** [comment]
Assuming all was fine with this fix, closing
